src/datasources/balances-api/valk-balances-api.service.ts

```diff
@@ -0,0 +1,143 @@
+      'balances.providers.valk.apiKey',
+    );
+    this.baseUri = this.configurationService.getOrThrow<string>(
+      'balances.providers.valk.baseUri',
+    );
+    this.defaultExpirationTimeInSeconds =
+      this.configurationService.getOrThrow<number>(
+        'expirationTimeInSeconds.default',
+      );
+    this.defaultNotFoundExpirationTimeSeconds =
+      this.configurationService.getOrThrow<number>(
+        'expirationTimeInSeconds.notFound.default',
+      );
+    this.chainsConfiguration = this.configurationService.getOrThrow<
+      Record<number, ChainAttributes>
+    >('balances.providers.valk.chains');
+  }
+
+  async getBalances(args: {
+    chainId: string;
+    safeAddress: string;
+    fiatCode: string;
+  }): Promise<Balance[]> {
+    try {
+      const cacheDir = CacheRouter.getValkBalancesCacheDir(args);
+      const chainName = this.getChainName(args.chainId);
+      const url = `${this.baseUri}/balances/token/${args.safeAddress}?chain=${chainName}`;
+      const valkBalances = await this.dataSource.get<ValkBalance[]>({
+        cacheDir,
+        url,
+        notFoundExpireTimeSeconds: this.defaultNotFoundExpirationTimeSeconds,
+        networkRequest: { headers: { Authorization: `${this.apiKey}` } },
+        expireTimeSeconds: this.defaultExpirationTimeInSeconds,
+      });
+      return this.mapBalances(valkBalances, args.fiatCode);
+    } catch (error) {
+      throw new DataSourceError(
+        `Error getting ${args.safeAddress} balances from provider: ${asError(error).message}}`,
+      );
+    }
+  }
+
+  mapBalances(valkBalances: ValkBalance[], fiatCode: string): Balance[] {
+    return valkBalances.map((valkBalance) => {
+      const price = valkBalance.prices[fiatCode.toUpperCase()] ?? null;
+      const balance = getNumberString(valkBalance.balance);
+      const fiatBalance = getNumberString(
+        (valkBalance.balance / Math.pow(10, valkBalance.decimals)) * price,
+      );
+      const fiatConversion = getNumberString(price);
+
+      // Valk returns a string representing the native coin (e.g.: 'eth') as
+      // token_address for native coins balances. An hex is returned for ERC20 tokens.
+      if (isHex(valkBalance.token_address)) {
+        return {
+          tokenAddress: valkBalance.token_address,
+          token: {
+            name: valkBalance.name,
+            symbol: valkBalance.symbol,
+            decimals: valkBalance.decimals,
+            logoUri: valkBalance.logo ?? '',
+          },
+          balance,
+          fiatBalance,
+          fiatConversion,
+        } as Erc20Balance & {
+          fiatBalance: string | null;
+          fiatConversion: string | null;
+        };
+      } else {
+        return {
+          tokenAddress: null,
+          token: null,
+          balance,
+          fiatBalance,
+          fiatConversion,
+        } as NativeBalance & {
+          fiatBalance: string | null;
+          fiatConversion: string | null;
+        };
+      }
```

Comment:
Nit: I'd separate the `Erc20Balance`/`NativeBalance` objects into their own variables then cast to `Balance` with the required values for better type safety.

```ts
const erc20Balance: Erc20Balance = {
  ...
}
return {
  ...erc20Balance,
  ...
} as Balance
```

Reply:
I think I found an implementation that is simpler than the one I initially did (just one return point) and also rolls back the need to export `Erc20Balance` and `NativeBalance` (increasing encapsulation). You can check it out in 

If you find this implementation a bit convoluted or you think it provides less type safety, I'd be happy to implement it as you proposed! 🙂

Reply:
Sorry for introducing more confusion, but I also moved up the public function (`getBalances`, `clearBalances`) and I made explicitly private the private ones (including the `_` prefix) :) [eff163c](https://github.com/safe-global/safe-client-gateway/pull/1062/commits/eff163cb9d44f9d482eb5300b4d0024009c0cf57)